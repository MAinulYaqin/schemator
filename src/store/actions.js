export const SET_TABLES = "SET_TABLES";

/**
 * Get action to set tables
 *
 * @param {object[]} tables Specified tables
 * @param {string} tables[].id Table ID
 * @param {string} tables[].name Table name
 * @param {number} tables[].timestamp Table creation date timestamp
 * @param {object} tables[].position Table position
 * @param {number} tables[].position.x Position X
 * @param {number} tables[].position.y Position Y
 * @param {object[]} tables[].fields Table field list
 * @param {string} tables[].fields[].id Field ID
 * @param {string} tables[].fields[].name Field name
 * @param {string} tables[].fields[].type Field type
 * @returns {object} Action
 */
export const setTables = tables => ({
  type: SET_TABLES,
  payload: tables
});
