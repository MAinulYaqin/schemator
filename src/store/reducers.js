import MdInfo from "react-icons/lib/md/info";

import {
  SET_TABLES,
  CLEAR_TABLES,
  ADD_TABLE,
  REMOVE_TABLE,
  UPDATE_TABLE,
  SET_FIELDS,
  CLEAR_FIELDS,
  ADD_FIELD,
  REMOVE_FIELD,
  UPDATE_FIELD,
  SET_RELATIONS,
  CLEAR_RELATIONS,
  ADD_RELATION,
  REMOVE_RELATION,
  UPDATE_RELATION,
  SET_PROJECT,
  SET_ALERT
} from "./actions";

export const tables = (state = [], action) => {
  const { type, payload } = action;

  switch (type) {
    case SET_TABLES:
      return payload;

    case CLEAR_TABLES:
      return [];

    case ADD_TABLE:
      return [...state, payload];

    case REMOVE_TABLE:
      return state.filter(item => item.id !== payload);

    case UPDATE_TABLE:
      return state.map(item => {
        if (item.id === payload.id) {
          return {
            ...item,
            ...payload.data
          };
        }

        return item;
      });

    default:
      return state;
  }
};

export const fields = (state = [], action) => {
  const { type, payload } = action;

  switch (type) {
    case SET_FIELDS:
      return payload;

    case CLEAR_FIELDS:
      return [];

    case ADD_FIELD:
      return [...state, payload];

    case REMOVE_FIELD:
      return state.filter(item => item.id !== payload);

    case UPDATE_FIELD:
      return state.map(item => {
        if (item.id === payload.id) {
          return {
            ...item,
            ...payload.data
          };
        }

        return item;
      });

    default:
      return state;
  }
};

export const relations = (state = [], action) => {
  const { type, payload } = action;

  switch (type) {
    case SET_RELATIONS:
      return payload;

    case CLEAR_RELATIONS:
      return [];

    case ADD_RELATION:
      return [...state, payload];

    case REMOVE_RELATION:
      return state.filter(item => item.id !== payload);

    case UPDATE_RELATION:
      return state.map(item => {
        if (item.id === payload.id) {
          return {
            ...item,
            ...payload.data
          };
        }

        return item;
      });

    default:
      return state;
  }
};

export const project = (state = null, action) => {
  const { type, payload } = action;

  switch (type) {
    case SET_PROJECT:
      return payload;

    default:
      return state;
  }
};

export const alert = (
  state = {
    isOpen: false,
    message: "My message",
    icon: MdInfo,
    iconColor: "black"
  },
  action
) => {
  const { type, payload } = action;

  switch (type) {
    case SET_ALERT:
      return {
        ...state,
        ...payload
      };

    default:
      return state;
  }
};
