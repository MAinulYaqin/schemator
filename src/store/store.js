import { combineReducers, createStore } from "redux";

import project from "./reducers/project";
import tables from "./reducers/tables";
import fields from "./reducers/fields";
import relations from "./reducers/relations";
import recentProjects from "./reducers/recentProjects";
import page from "./reducers/page";

const rootReducer = combineReducers({
  project,
  tables,
  fields,
  relations,
  recentProjects,
  page
});

const store = createStore(
  rootReducer,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
);

export default store;
