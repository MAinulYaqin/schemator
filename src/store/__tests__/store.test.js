import store from "../store";

describe("store", () => {
  it("should not error", () => {
    const state = store.getState();

    expect(state).toMatchSnapshot();
  });
});
