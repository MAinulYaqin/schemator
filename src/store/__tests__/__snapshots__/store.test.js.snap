// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store should not error 1`] = `
Object {
  "extension": null,
  "extensions": Array [],
  "fields": Array [],
  "page": "workarea",
  "project": null,
  "recentProjects": Array [],
  "relations": Array [],
  "tables": Array [],
}
`;
